Siren system

12gabriel3 opened the pull request:
This adds a new system for emergency vehicles.

See description in https://github.com/12gabriel3/EveronLife/blob/siren/docs/siren_system/index.md

12gabriel3 committed:
Add siren

12gabriel3 committed:
Fix merge conflicts

12gabriel3 committed:
Add emergency lights

12gabriel3 committed:
First working example

12gabriel3 committed:
Add basic light animations

12gabriel3 committed:
Add user configuration

12gabriel3 committed:
Add basic prefabs

12gabriel3 committed:
Merge branch 'main' into siren

12gabriel3 committed:
Fix character rocking back and forth in stopped vehicle

12gabriel3 committed:
Set RigidBody of police vehicle to inactive

12gabriel3 committed:
Cleanup

12gabriel3 committed:
Group similar lights

12gabriel3 committed:
Add glow to knob

12gabriel3 committed:
Remove old config file

12gabriel3 committed:
Increase emissive lv

12gabriel3 committed:
Improve UX

12gabriel3 committed:
Remove volume light from blue LightEntity

12gabriel3 committed:
Add config file for siren customization

12gabriel3 committed:
Add custom titles

12gabriel3 committed:
Add descriptions

12gabriel3 committed:
Fix knob white color during day

12gabriel3 committed:
Add faster siren

12gabriel3 committed:
Add faster siren as an option

12gabriel3 committed:
Change base configuration

12gabriel3 committed:
Change audio project to new file instead of override

12gabriel3 committed:
Reorganize assets

12gabriel3 committed:
Audio cleanup

12gabriel3 committed:
Add example prefabs

12gabriel3 committed:
Match red and blue light

12gabriel3 committed:
Add mp support

12gabriel3 committed:
Change spawn to police prefab

12gabriel3 committed:
Add basic siren configs

12gabriel3 committed:
Change modes names and order

12gabriel3 committed:
Add docs

12gabriel3 committed:
Commented the code

12gabriel3 committed:
Feature: adaptive LightEntity LV

12gabriel3 committed:
Fix TimeAndWeatherManager time and environment not matching

12gabriel3 committed:
Adjust sound and light intensity

12gabriel3 committed:
Merge branch 'main' into siren

12gabriel3 committed:
Fix light and siren not synching for JIPs

12gabriel3 committed:
Add keybinds for siren and sound for mode change

12gabriel3 committed:
Change siren when honking on cruise mode to fastest

Arkensor commented on docs/siren_system/further_modding.md:
```diff
@@ -0,0 +1,164 @@
+
+Make a model and set up the procedural animation as usual. The knob prefab must have:
+
+- A `MeshObject` with a collider
+- A `RigidBody` set to `AlwaysActive`
+- A `ProcAnimComponent`
+- A `SignalsManagerComponent`
+- A `Hierarchy`
+- An `EL_SirenKnobComponent`
```
```suggestion
- `MeshObject` with a collider
- `RigidBody` set to `AlwaysActive`
- `ProcAnimComponent`
- `SignalsManagerComponent`
- `Hierarchy`
- `EL_SirenKnobComponent`
```
Solves the problem :)

Arkensor commented on docs/siren_system/index.md:
```diff
@@ -0,0 +1,39 @@
+
+# Siren system
+
+The siren system introduces selectable light animations and sirens to [Arma Reforger](https://store.steampowered.com/app/1874880).
```
```suggestion
The siren system introduces selectable light animations and sirens to Arma Reforger.
```

Arkensor commented on docs/siren_system/index.md:
```diff
@@ -0,0 +1,39 @@
+
+# Siren system
```
Unsure if you should list all the index of the subfiles again here.

12gabriel3 replied:
I ended up taking this approach because wanted to make sure people would have an overview of what is advanced usage and what is further modding. 
> I want to change light animations
Go to the index, it's there, under advanced usage 
If I linked only the subfiles, people would have to wonder which file contains what they are looking for.

This makes the index kind of cluttered, though. So either approach is good for me.

Even better would be to make a FAQ

Arkensor replied:
Well If I want to change the light animations I will look at the 3 available pages and their short description and see which one appears to be what I need. I think the index.md should contain not more than than. Especially in regards of if something is changed in the sub change (like a heading is added) then you also always need to edit the main index ...

Arkensor reviewed (commented):
I tried to go through everything, but it is a lot. I will need to do at least one other review pass once all the notes above are discussed to spot anything in the "final" design.

# General notes
- Please go through the files and make sure you did not include any working temp files. I am not sure if there are some unused materials, textures, or models in the files right now. Double-check all files listed in the commit, please.
- Some names are very specific now about M151A2 for example. If the prefab could be put on top of other vehicles too (which they should) then the different siren light prefabs should have generic names for the prefabs.
- We have generally preferred not to configure any of the vanilla vehicles, but I think there are so few and the prefabs take so few bytes, that we can just setup siren lights for them. But we can not directly add it to the main prefab, instead, we will need to make a police variant of the vehicles or else the civilian one from the shop had them as well :)
- I mainly looked for formatting etc right now and did not dive deep into the code and optimizations yet. When everything else is sorted out I will see if I find something, but judged from reading it once it looks well thought out.
- Some of the assets are not pascal case. e.g. `siren.wav` -> `Siren.wav`, `sirenFastest.wav` -> `SirenFastest.wav`

# Technical design 
- I strongly suggest to rename `EL_LightComponent` to `EL_SirenLightComponent`. That fits the other "siren" prefixed or suffixed names and strongly associates it with the feature you are building. Just light component would be a general light we use for all sorts of things - including unrelated to sirens.
- Looking at the end result gifs, especially with the led lights the siren configuration seems to make cool things possible. One thing I do not understand is why there is turn lights on/off vs toggle. Why not only have turn on and off? Can you give a use case example where you want to toggle? I am not against having it, I am just curious about why it was added, you must have had a good reason to do all the extra work.
- I think it would help if the `Name` was renamed to `Lights` or `Light names` for the config entry.
- I think the configurable loop entry was a good idea. Most will be from the beginning and infinite, but it opens up the possibility for a "startup" animation or a light sequence that only runs once. I might give some better naming to the attribute to say `Start From` instead of `Steps`? 

# Gameplay design 
- The general looks are very nice, and I love the animations from the led lights especially. One thing I said on discord before that continues to "upset" me a little bit is the black empty void when you have rotating lights. The vehicle has a radius of nothing around it and the shell itself is not light up at all. In reality, even though the light is spinning the plastic around it would scatter the light to both light up the hull of the light as well as diffuse onto the surrounding vehicle there. Not too much but not full darkness. Not sure what is possible without adding too many more light sources. Maybe reduce the min range of the rotating spotlight and have that be a "bit unrealistic" than the clear-cut black void.
- The horn button as a shortcut works, but it conflicts a tiny bit with just quickly touching the horn. You can still do both sure, but I accidentally turned on the lights a few times during testing. I also preferred that while having a siren mode active the horn was remapped to make the siren go faster for the duration of the horn press. I think bringing back that design and adding re-mappable keybinds for siren modes that we put on the f keys by default *could* be nicer. Let us discuss that and see what others think as well.

# Documentation
- I am not sure if you need to put the full index of each sub page into the main index.md, or if you just link to subpages with a 2-3 sentences description of what is described there.
- When viewing on desktop some images are wrapped by text, maybe you want to move the text before or after the image, or else it is a big bugged - not a huge issue though :) 
![image](https://user-images.githubusercontent.com/8494013/178438888-3686041d-1c60-4f21-a094-1ea128b2f7f1.png)

12gabriel3 committed:
Update docs/siren_system/further_modding.md

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 committed:
Update docs/siren_system/index.md

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 committed:
Rename EL_LightComponent to EL_SirenLightComponent

12gabriel3 committed:
Rename M151A1 instances to M151A2

12gabriel3 committed:
Cleanup

12gabriel3 committed:
Rename sounds to pascal case

12gabriel3 committed:
Cleanup

12gabriel3 committed:
Merge branch 'siren' of github.com:12gabriel3/EveronLife into siren

12gabriel3 commented on src/Scripts/Game/Components/Vehicle/EL_SirenManagerComponent.c:
```diff
@@ -0,0 +1,370 @@
+	};
+	
+	// Available sounds
+	static protected const ref ParamEnumArray SIREN_SOUNDS = 
```
Previous array is not protected because `EL_SirenModeAction` needs to access the available modes, while the sounds are only selected in `EL_SirenMode` class

12gabriel3 committed:
Update docs/siren_system/further_modding.md

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 commented on src/Scripts/Game/Components/Lights/EL_LightComponent.c:
```diff
@@ -0,0 +1,284 @@
+	
+	[Attribute("30", UIWidgets.EditBox, "Radius of the light effect")]
+	private float m_fRadius;
+	
```
They were private on accident

12gabriel3 commented on src/Scripts/Game/Global/EL_Attributes.c:
```diff
@@ -0,0 +1,42 @@
+
+/**
+	\brief Attribute for setting any enum property as custom title.
+	@code
+	[BaseContainerProps(), SCR_BaseContainerCustomTitleEnum(EEditorMode, "m_Mode")]
+	class TestConfigClass
+	{
+		[Attribute()]
+		private EEditorMode m_Mode;
+	}
+	@endcode
+*/
+class EL_BaseContainerCustomTitleEnumReadable : BaseContainerCustomTitle
```
This is a modification of `BaseContainerCustomTitleEnum` in `Atributes.c`.
The original class would print enums as text just like you typed it.
With this one I'm able to display the entries to the user with the tile `Turn lights off` as opposed to `TURN_LIGHTS_OFF`

12gabriel3 committed:
Code formatting

12gabriel3 committed:
Cleanup debug world

12gabriel3 committed:
Update docs/siren_system/further_modding.md

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 committed:
Delete not needed sound files

12gabriel3 committed:
Merge branch 'siren' of github.com:12gabriel3/EveronLife into siren

12gabriel3 commented on src/Sounds/Siren/Vehicles_Siren.afm:
```diff
@@ -0,0 +1,2 @@
+AudioOutputStateResClass {
+}
```
I think it was auto generated, we don't need it

12gabriel3 commented on src/Sounds/FinalMix.afm.meta:
```diff
@@ -0,0 +1,15 @@
+MetaFileClass {
+ Name "{B764D803219C775F}Sounds/FinalMix.afm"
```
I copy and pasted the horn audio project to make the siren sounds. It did not work for some reason and I had to override some sounds files. 
But now I used another method to duplicate the horn audio project. I tested it and we don't need this file anymore.

12gabriel3 committed:
Move EL_BaseContainerCustomTitleEnumReadable into its own file

12gabriel3 commented on src/Scripts/Game/Components/Lights/EL_LightComponent.c:
```diff
@@ -0,0 +1,284 @@
+												m_currentLV,
+												m_vLightOffset,
+												m_vLightConeDirection);
+		m_LightEntity.SetConeAngle(m_fConeAngle);
```
I think the main problem here is that the Script Editor in the Workbench uses tabs instead of spaces.

12gabriel3 committed:
Rename m_Steps to m_StartFrom

12gabriel3 commented on src/Scripts/Game/Components/Vehicle/EL_SirenManagerComponent.c:
```diff
@@ -0,0 +1,370 @@
+class EL_SirenMode
+{
+	// Available modes to the user
+	static const ref ParamEnumArray MODE_NAMES =
```
Removing the static keyword gives me compiling errors

12gabriel3 replied:
I get the error 
```Trying to access non-static member 'MODE_NAMES' from static method '#CreateAttributesEL_SirenMode'```

12gabriel3 commented:
I modified the code like you suggested, and answering some of your points:

* The prefab and model named after the M151A2 is a light pole made specifically for it. You can still make different mounts for different vehicles and just add the actual light casing (which has a generic name) to it.

* The toggle option was added for the convenience of the person configuring the animation. As you control more lights with complex animations, it becomes very confusing very quickly. The toggle entry makes it *much* easier to understand the animation flow.

* About the circular shadow, it really is just a characteristic of this type of emergency light. You can see [here](https://www.youtube.com/watch?v=oRiocQQ1SD8) and [here](https://www.youtube.com/watch?v=le1Cy9wLneQ) that even with a very close white wall reflecting the light back, near the middle there's shadow. I think anything we do with this specific model would feel even more off. And feeling off is a matter of opinion, at the end, server owners can customize it the way they like.

* Pressing the horn still makes the siren go faster. I'm not sure what you are referring to. Either way it can be configured through a simple combo box.

* The shortcuts were designed to be console friendly. It might not be the ideal scenario for PC because, as you noticed, it adds a slight delay to the horn. We have many options here and we should discuss, but if we choose to keep the control scheme (which I think is good for console), one option that might be worth doing is adding a separate control context to police cars, so that horns on other vehicles aren't affected, if that is possible at all.

12gabriel3 committed:
Make MODE_NAMES and SIREN_SOUNDS static

Arkensor commented:
> About the circular shadow, it really is just a characteristic of this type of emergency light. You can see [here](https://www.youtube.com/watch?v=oRiocQQ1SD8) and [here](https://www.youtube.com/watch?v=le1Cy9wLneQ) that even with a very close white wall reflecting the light back, near the middle there's shadow. I think anything we do with this specific model would feel even more off. And feeling off is a matter of opinion, at the end, server owners can customize it the way they like.

Yes, there is a tiny bit of that but that is no comparison to having a 4-meter void around the vehicle. There should be light much closer. The light normally does not travel just 100% straight forward from them, because of the mentioned refraction from the casing. That is one reason why the casings are not clear, they are intended to scatter parts of the light so the vehicle is lit up as well. Yes, server owners *can* customize it but many won't. Can we please try and see what is possible in bringing it closer? Try producing no gap at all and let's see how that looks.

> Pressing the horn still makes the siren go faster. I'm not sure what you are referring to. Either way it can be configured through a simple combo box.

When I played around with it, it switched the mode permanently, what I wanted is to hold it for a bit where the siren increases speed and when I release it goes back to the current normal level. Maybe I miss-clicked.

> The shortcuts were designed to be console friendly. It might not be the ideal scenario for PC because, as you noticed, it adds a slight delay to the horn. We have many options here and we should discuss, but if we choose to keep the control scheme (which I think is good for console), one option that might be worth doing is adding a separate control context to police cars, so that horns on other vehicles aren't affected, if that is possible at all.

Ah yes, consoles. Right, that might be a good reason to use this shortcut as they do not have F keys. We *could* still add shortcut keybinds for PC and have a setting if horn is "abused" for it, but I guess we should keep it simple and in parity for both platforms.

Arkensor commented:
I have noticed there is a `{F649585ABB3706C4}Prefabs/Vehicles/Wheeled/M151A2/M151A2.et` prefab that turns the vanilla vehicle blue? I think this might be a leftover since you made a police base already?

Arkensor commented:
Parts of the LED bar are invisible when turned on depending on the view angle.
![image](https://user-images.githubusercontent.com/8494013/178674351-8b6a6f7f-2396-4756-aa21-85483a500827.png)

12gabriel3 committed:
Optimize replication

12gabriel3 committed:
Fix LightEntities not spawning for JIPs

12gabriel3 committed:
Fix lights not starting on default mode

12gabriel3 committed:
Fix switching siren mode through keybind

Fixes the keybind switching the mode on every police car and it not being broadcast to other clients

Arkensor commented on src/Scripts/Game/Components/Vehicle/EL_SirenManagerComponent.c:
```diff
@@ -187,6 +188,8 @@ class EL_SirenMode
 	};
 	
+	// Number of modes available, used to reduce network traffic. Increase in powers of 2
+	static const int N_MODES = 8;
```
Does this need to be hard coded? Why not just array.Count()? that could likely be made const as well if the array is const so it's only ever evaluated once.

e.g. `const int count = EL_SirenMode.MODE_NAMES.Count()`.

And I think the WriteIntRange is clever enough to know how many bits to use even for odd numbers.

12gabriel3 committed:
Fix hardcoded number of available modes

12gabriel3 committed:
Increase max emissive multiplier

12gabriel3 committed:
Rename materials and add lods to EmergencyLightRotaryBase

12gabriel3 committed:
Rename materials, add lods and colliders to LightSingleRotary and LightSingleRotaryBuiltIn

12gabriel3 committed:
Rename materials, add colliders and LODs to M151A2_lightsMount

12gabriel3 committed:
Add rigid bodies

12gabriel3 committed:
Add EmergencyLightBuiltIn prefab

12gabriel3 committed:
Add new light bar

12gabriel3 committed:
Change spawned vehicle in debug world to newly added lightbar

12gabriel3 committed:
Remove M151A2 material override and fix searchlight under vehicle bug

12gabriel3 committed:
Add lightbar texture

12gabriel3 committed:
Change search light angle

12gabriel3 committed:
Add light bar colliders

12gabriel3 committed:
Add M151A2 light bar mount

12gabriel3 committed:
Fix black circle

12gabriel3 committed:
Change emissive color to a more realistic yellow

12gabriel3 commented on src/Scripts/Game/Components/Lights/EL_LightComponent.c:
```diff
@@ -0,0 +1,284 @@
+	protected vector m_EmissiveColor;
+	
+	[Attribute("2", uiwidget: UIWidgets.Slider, params: "0 100")]
+	protected int m_EmissiveMultiplier;
```
These are for every instance, as someone could want different emissive color for the same asset. ie blue on the left and red on the right. The color might also change the percieved brightness which can be adjusted individually

12gabriel3 committed:
Change index to only main pages

12gabriel3 committed:
Update src/Scripts/Game/Components/Vehicle/EL_SirenManagerComponent.c

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 committed:
Formatting

12gabriel3 committed:
Fix invisible texture on LED bar

12gabriel3 committed:
Reorganised directory structure

12gabriel3 committed:
Fix knob wrong position on mode 3

12gabriel3 committed:
Fix prefab searching for component with old name

Arkensor reviewed (changes requested):
Before getting into the nit-picky details of this review, I want to express my respect for this very well-built feature. It is quite complex and we went through many iterations, but now we are approaching "perfect" territory. I think that after this review has been applied, the feature is not only ready for merge but is unlikely to ever require any significant patches (unless somebody wants to do some minor enhancements).

# Assets
- `M151A2_lightsMount.fbx` -> `M151A2_LightMount.fbx` to match `M151A2_LightMount.emat`?
- `M151A2_BarMount.fbx` vs `M151A2_LightBarMount.emat` missmatch again
- `Assets\Vehicles\Wheeled\M151A2\Data\Mount.emat` needed or dev leftover?

- Do you want to keep the folder name `EmergencyLights`? Or do you want to match it with scripts to maybe be `SirenLights`?

- Do you think we should maybe move "attachments" like these into a dedicated folder `Assets\Vehicles\Attachments` so on that level we have `Attachments`, `Wheeled`, `Tracked`, `Planes` etc in the distant future? We can move siren light stuff there first in this PR and move licenses plates in the update under construction in [enhancement/license_plates_cleanup](https://github.com/EveronLife/EveronLife/tree/enhancement/license_plates_cleanup)

- Can we make police variants of all existing vanilla vehicles with the basic blue police material? applied too? I think sooner or later we need them anyway. There are only a few vehicles, so that should be alright? Maybe we can add them in a way to only the base variants of the vehicles so they do not clip with any of the vanilla subvariants?
- The above would also "fix" the `M923A1` truck only consisting of the floating stuff. I can see that other parts are often floating as well. I am not sure how the setup for this very modular truck is supposed to be used and if or how we could make a re-useable police variant for them.

- `M151A2/M151A2.et` already contains the knob action, even if the knob is not present.
- The base variant of something is per convention in AR with _Base, so it would be `M151A2/M151A2Police_Base.et` or likely even better `M151A2/M151A2_Police_Base.et`. As we will have `M151A2_Medics_xxx` etc etc. The game uses _ to separate these distinct categories :) 

- The very cool roof mount + bar 2 seems to have some issue where part of the lights are not actually in the rotary2 object
![image](https://user-images.githubusercontent.com/8494013/179918750-38641f4f-9940-44fd-8d11-ae186fe5a857.png)

- Any idea why the game shows such a tall shape around the pole variant?
![image](https://user-images.githubusercontent.com/8494013/179920304-1d19b788-34db-481f-9061-90aeec26be2e.png)

- `Create entity @"ENTITY:169" ('GameEntity') at <118.016335 1.924212 152.726440>
 RESOURCES (E): Can't remap 'Assets/Vehicles/EmergencyLights/EmergencyLightRotaryBase.xob' material from 'Lights' to 'Assets/Vehicles/Wheeled/Attachments/Lights/Emergency/Base/Data/Lights.emat'. Source material do not exist`

# Behavior
- I noticed that because we now override the horn action if you are in a vehicle that does not have siren lights you can no longer "quick tap" the horn and need to hold for a bit. Not a huge issue, and I don't see a way around it unless we could selectively have different actions context for the horn depending on what kind of vehicle it is. Or maybe we can manually activate the horn quickly on the "one tap" trigger that is mainly used to quickly enable the siren mode. So if sirens are not there it functions in the original role of playing the horn for a short amount of time.

# Scripts - Part 1
- Not your fault, I did not yet have time to clean it up, but the `Scripts/Game/Components` folder should not exist at all. We are not doing any "one place per x" but instead vertical slices for features. So You just need to move the files to (for now) `Scripts/Vehicle/SirenLights` maybe? Luckily scripts can be moved all they want with no need to adjust anything anywhere. I will be moving the license plate stuff there in the before-mentioned open working branch for it as well as remove the trader once that system is finished soon :)
- Regarding some formatting/bracket placing rules I kicked off [this discussion](https://github.com/EveronLife/EveronLife/discussions/51#discussioncomment-3186833) after reading the code. Depending on the outcome We might want to do some minor code adjustments
... to be expanded in Part 2 of the review later

12gabriel3 commented on src/Scripts/Game/Core/EL_Utils.c:
```diff
@@ -11,4 +11,53 @@ class EL_Utils
+		vector mat[4];
+		IEntity child = GetGame().SpawnEntityPrefab(resource, parent.GetWorld());
+		
+		parent.AddChild(child, parent.GetBoneIndex(pivot.GetName()));
```
It works. I used it to attach the light entity to the spinning emissive

12gabriel3 commented on src/Scripts/Game/Core/EL_BaseContainerCustomTitleEnumReadable.c:
```diff
@@ -0,0 +1,41 @@
+		{
+			return false;
+		}
+		string _title = typename.EnumToString(m_EnumType, enumValue);
+		string firstLetter = _title[0];
+		firstLetter.ToUpper();
+		_title.Replace("_", " ");
+		_title.ToLower();
+		_title = firstLetter + _title.Substring(1, _title.Length() - 1);
+		title = string.Format(m_sFormat, _title);
```
I blame the devs, I copied the names from them

12gabriel3 committed:
Move scripts to different folder

12gabriel3 committed:
Make member variables protected instead of private

12gabriel3 committed:
Change spawned vehicle to vanilla one

12gabriel3 committed:
Fix null pointer on OnPostInit()

12gabriel3 committed:
Rename assets to match materials and cleanup unused materials

12gabriel3 committed:
Add rotations to bones

This allows to attach children to the pivot points directly, without
adjusting the rotation manually. Should make it easier to modify in the
future

12gabriel3 committed:
Add pivot point for the light bar

12gabriel3 committed:
Move SirenModeActions to the Knob

This allows to create new vehicles by only adding the knob instead of
creating action contexts and actions for each one.

12gabriel3 committed:
Remove leftover actions from base M151A2 prefab

12gabriel3 committed:
Fix animation not working after changing loop attibute name

12gabriel3 committed:
Move search light emissive to light bar

Slots allow you to register the lights from children to the parent.
This makes it possible to have the emissive surface on the lightbar,
making it much easier to set up on new vehicles

12gabriel3 committed:
Create new prefabs reflecting changes

12gabriel3 committed:
Fix null pointer and remove unneeded sound event

12gabriel3 commented on src/Scripts/Game/Core/EL_BaseContainerCustomTitleEnumReadable.c:
```diff
@@ -0,0 +1,41 @@
+{
+	private typename m_EnumType;
+	private string m_PropertyName;
+	private string m_sFormat;
```
Done in https://github.com/EveronLife/EveronLife/pull/142/commits/5a65476a08353f052106aef1c23bd14a60481ba9

12gabriel3 committed:
Update src/Scripts/Game/Core/EL_BaseContainerCustomTitleEnumReadable.c

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 committed:
Update src/Scripts/Game/Core/EL_Utils.c

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 committed:
Update src/Scripts/Game/Core/EL_Utils.c

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 committed:
Update src/Scripts/Game/Core/EL_Utils.c

Co-authored-by: Arkensor <arkensor@gmail.com>

12gabriel3 commented on src/Worlds/DebugWorld/DebugWorld_Spawnconfig.layer:
```diff
@@ -1,7 +1,7 @@
  coords 110.036 0.001 142.458
  angleY 90
- m_rnPrefab "{F649585ABB3706C4}Prefabs/Vehicles/Wheeled/M151A2/M151A2.et"
+ m_rnPrefab "{B25F00A6525B4221}Prefabs/Vehicles/Wheeled/M151A2/M151A2PoliceBar2.et"
```
Done in https://github.com/EveronLife/EveronLife/pull/142/commits/8640b539bb5d0a28c7bae6c9cbb14d345b065ed5

12gabriel3 committed:
Merge branch 'main' into siren

12gabriel3 committed:
Delete old lower case folder

12gabriel3 committed:
Rename materials and prefabs folders

12gabriel3 commented:
* Renamed materials that don't match the model name
* Renamed the asset folder to SirenLights
* Let us discuss a final folder structure because you need to move each folder individually as the workbench doesn't update file references otherwise. I think moving things to an `Attachments` folder would be best done this way:
```
├───Attachments                                    # Attachments not specific to any vehicle
│   ├───LicensePlate
│   └───SirenLights
├───Tracked
│   ├───Attachments                              # Attachments specific to tracked vehicles
│   │   ├───LicensePlate
│   │   └───SirenLights
│   ├───OtherTank
│   │   └───Attachments                        # Attachments specific to 'OtherTank'
│   │       ├───LicensePlate
│   │       └───SirenLights
│   └───Tank1                                         
│       └───Attachments          
│           ├───LicensePlate
│           └───SirenLights
└───Wheeled
    ├───Attachments
    │   ├───LicensePlate
    │   └───SirenLights
    ├───M151A2
    │   └───Attachments
    │       ├───LicensePlate
    │       └───SirenLights
    └───OtherCar
        └───Attachments
            ├───LicensePlate
            └───SirenLights
```
* Renamed the prefabs to match the convention
* The emissive part not being part of the rest of the LightBar wasn't an issue, that's how emissive surfaces work, they need to be part of the vehicle. I added them `BaseLightManagerComponent`, because since they are not animated or rotary, the vanilla vehicle light system is enough. Nevertheless, I found a way of adding the emissives to the Light Bar and transfer them to the vehicle through Slots. This makes the setup more convenient.
* Im not sure about the tall shape, but I remade the prefabs and it fixed itself.
* Retexturing all vehicles to blue is out of the scope I intended for this PR, because someone else is already retexturing vehicles for police. Unless they contribute to this PR, I think textures should stay as is.

Arkensor commented:
> Let us discuss a final folder structure because you need to move each folder individually as the workbench doesn't update file references otherwise. I think moving things to an Attachments folder would be best done this way:

I agree. That structure looks very logical to me.

> Retexturing all vehicles to blue is out of the scope I intended for this PR, because someone else is already retexturing vehicles for police. Unless they contribute to this PR, I think textures should stay as is.

I thought it maybe was "just" editing an RGB value somewhere to turn them blue. Sure we can leave that to a separate PR later

12gabriel3 committed:
Merge branch 'main' into siren

12gabriel3 committed:
Fix siren not working on 9.6

dan-schwerner commented:
hey @Arkensor @12gabriel3  any chance this can be merged to main? massive chunk of work would be a massive shame if it was lost
